qpid-jms-client/src/main/java/org/apache/qpid/jms/transports/netty/NettyEventLoopGroupPoolFactory.java

```diff
@@ -0,0 +1,253 @@
+                  return true;
+               }
+            }
+         }
+
+         @Override
+         public EventLoopGroup ref() {
+            if (refCnt.get() == 0) {
+               throw new IllegalStateException("the event loop group cannot be reused");
+            }
+            return group;
+         }
+
+         @Override
+         public void close() {
+            while (true) {
+               final int currValue = refCnt.get();
+               if (currValue == 0) {
+                  return;
+               }
+               if (refCnt.compareAndSet(currValue, currValue - 1)) {
+                  if (currValue == 1) {
+                     // if a racing thread has borrowed this lease, it would try already to set it to a new value:
+                     // this one is a best effort cleanup to help GC
+                     sharedEventLoopGroup.compareAndSet(this, null);
+                     Future<?> fut = group.shutdownGracefully(0, SHUTDOWN_TIMEOUT, TimeUnit.MILLISECONDS);
+                     if (!fut.awaitUninterruptibly(2 * SHUTDOWN_TIMEOUT)) {
+                        LOG.trace("Channel group shutdown failed to complete in allotted time");
+                     }
+                  }
+                  return;
+               }
+            }
+         }
+      }
+
+      private final AtomicReference<SharedGroupRef> sharedEventLoopGroup = new AtomicReference<>(null);
+
+      @Override
+      public Ref<EventLoopGroup> sharedGroupWith(final int threads) {
+         if (threads <= 0) {
+            throw new IllegalArgumentException("threads must be > 0");
+         }
+         while (true) {
+            SharedGroupRef sharedGroupRef = sharedEventLoopGroup.get();
+            if (sharedGroupRef != null && sharedGroupRef.retain()) {
+               validateSharedRef(sharedGroupRef, threads);
+               return new AtomicCloseableRef<>(sharedGroupRef);
+            }
+            synchronized (this) {
+               sharedGroupRef = sharedEventLoopGroup.get();
+               if (sharedGroupRef != null && sharedGroupRef.retain()) {
+                  validateSharedRef(sharedGroupRef, threads);
+                  return new AtomicCloseableRef<>(sharedGroupRef);
+               }
+               sharedGroupRef = new SharedGroupRef(createEventLoopGroup(threads, this, new QpidJMSThreadFactory("SharedNettyEventLoopGroup :(" + SHARED_NETTY_EVENT_LOOP_GROUP_SEQUENCE.incrementAndGet() + ")", true)), threads);
+               sharedEventLoopGroup.set(sharedGroupRef);
+               return new AtomicCloseableRef<>(sharedGroupRef);
+            }
+         }
+      }
+
+      private static void validateSharedRef(final SharedGroupRef sharedRef, final int threads) {
+         if (threads != sharedRef.getThreads()) {
+            try (SharedGroupRef ref = sharedRef) {
+               throw new IllegalArgumentException("threads cannot be different from " + sharedRef.getThreads());
+            }
+         }
+      }
```

Comment:
Doubt Tim will like this anymore than the prior throwing if the types differed. I dont particularly either. I think it either needs to handle it somehow (e.g distinct pools), or just proceed with whats there and warn if its less (or perhaps just different) than requested.

Unclear why the assignment is in the try, when its value then isnt used; did you mean to use ref instead of sharedRef in the exception message? Even if so, wouldnt that still be racey with the sharedRef use in the out enclosing if? Or is it so it closes it after throwing because its autoClosable? Whatever it is, if an update doesnt make it clearer, its probably worth a comment unless this is just removed.

Reply:
> Unclear why the assignment is in the try, when its value then isnt used

I've decided the "hard" approach where even checking the `getThreads` cannot happen if the ref isn't retained yet: autoclose it is necessary because of the previous `retain`

Reply:
Let me think how I can make it better to make Tim happy about it :)